library/src/hspec/stm/stm32f7/mal_hspec_stm32f7_spi.c

```diff
@@ -0,0 +1,431 @@
+        // Set unselected
+        if (MAL_SPI_SELECT_POLARITY_HIGH == init->select_polarity) {
+            mal_result = mal_gpio_set(&handle->select, false);
+        } else {
+            mal_result = mal_gpio_set(&handle->select, true);
+        }
+        if (MAL_ERROR_OK != mal_result) {
+            return mal_result;
+        }
+    }
+    // Reset interface
+    hal_result = HAL_SPI_DeInit(&handle->hal_spi_handle);
+    if (HAL_OK != hal_result) {
+        return MAL_ERROR_HARDWARE_INVALID;
+    }
+    // Initialize interface
+    handle->hal_spi_handle.Init.TIMode = SPI_TIMODE_DISABLE;
+    handle->hal_spi_handle.Init.CRCCalculation = SPI_CRCCALCULATION_DISABLE;
+    // Set full duplex mode
+    handle->hal_spi_handle.Init.Direction = SPI_DIRECTION_2LINES;
+    // Set as master
+    handle->hal_spi_handle.Init.Mode = SPI_MODE_MASTER;
+    // Set word size
+    if (init->data_size < MAL_SPI_DATA_4_BITS || init->data_size > MAL_SPI_DATA_16_BITS) {
+        return MAL_ERROR_HARDWARE_INVALID;
+    }
+    handle->hal_spi_handle.Init.DataSize = (uint16_t)(init->data_size - 1) << 8;
+    // Set clock idle polarity
+    if (MAL_SPI_CLK_IDLE_STATE_LOW == init->clk_idle_state) {
+        handle->hal_spi_handle.Init.CLKPolarity = SPI_POLARITY_LOW;
+    } else {
+        handle->hal_spi_handle.Init.CLKPolarity = SPI_POLARITY_HIGH;
+    }
+    // Set on which edge data should be valid
+    if (MAL_SPI_CLK_IDLE_STATE_LOW == init->clk_idle_state &&
+        MAL_SPI_DATA_LATCH_EDGE_RISING == init->latch_edge) {
+        handle->hal_spi_handle.Init.CLKPhase = SPI_PHASE_1EDGE;
+    } else if (MAL_SPI_CLK_IDLE_STATE_HIGH == init->clk_idle_state &&
+               MAL_SPI_DATA_LATCH_EDGE_FALLING == init->latch_edge) {
+        handle->hal_spi_handle.Init.CLKPhase = SPI_PHASE_1EDGE;
+    } else {
+        handle->hal_spi_handle.Init.CLKPhase = SPI_PHASE_2EDGE;
+    }
+    // Set select (NSS) control mode
+    switch (init->select_mode) {
+        case MAL_SPI_SELECT_MODE_HARDWARE:
+            // This MCU only supports low polarity in hardware mode
+            if (MAL_SPI_SELECT_POLARITY_HIGH == init->select_polarity) {
+                return MAL_ERROR_HARDWARE_INVALID;
+            }
+            handle->hal_spi_handle.Init.NSS = SPI_NSS_HARD_OUTPUT;
+            handle->hal_spi_handle.Init.NSSPMode = SPI_NSS_PULSE_ENABLE;
+            break;
+        case MAL_SPI_SELECT_MODE_SOFTWARE:
+        case MAL_SPI_SELECT_MODE_USER:
+        case MAL_SPI_SELECT_MODE_NONE:
+        default:
+            handle->hal_spi_handle.Init.NSS = SPI_NSS_SOFT;
+            handle->hal_spi_handle.Init.NSSPMode = SPI_NSS_PULSE_DISABLE;
+            break;
+    }
+    // Get SPI clock source
+    // Set the clock prescaler. The prescaler can only divide with powers of 2.
+    // The valid values range from 2 to 256. In other words 2^1 to 2^8. The
+    // prescaler value of the register is 0 based. This means you have to
+    // subtract 1 to the value to set to the register.
```

Comment:
Should this line say "add 1" instead of "substract 1"?

Reply:
Sorry after reading again a 3rd time, I now understand what you mean

Reply:
No, this is the explanation of if you do it on paper and try to compare your result with what is in the register. If you don't subtract 1, you will be off by 1.